File: src/js/4_BehavioralPatterns/6_Memento/memento.js
class Memento
{
  constructor(balance)
  {
    this.balance = balance;
  }
}

class BankAccount
{
  constructor(balance=0)
  {
    this.balance = balance;
  }

  deposit(amount)
  {
    this.balance += amount;
    return new Memento(this.balance);
  }

  restore(m)
  {
    this.balance = m.balance;
  }

  toString()
  {
    return `Balance: ${this.balance}`;
  }
}

let ba = new BankAccount(100);
let m1 = ba.deposit(50);
let m2 = ba.deposit(25);
console.log(ba.toString());

// restore to m1
ba.restore(m1);
console.log(ba.toString());

// restore to m2
ba.restore(m2);
console.log(ba.toString());

File: src/csharp/4_BehavioralPatterns/6_Memento/Memento.cs
using static System.Console;

namespace DotNetDesignPatternDemos.Behavioral.Memento
{
  public class Memento
  {
    public int Balance { get; }

    public Memento(int balance)
    {
      Balance = balance;
    }
  }

  public class BankAccount
  {
    private int balance;

    public BankAccount(int balance)
    {
      this.balance = balance;
    }

    public Memento Deposit(int amount)
    {
      balance += amount;
      return new Memento(balance);
    }

    public void Restore(Memento m)
    {
      balance = m.Balance;
    }

    public override string ToString()
    {
      return $"{nameof(balance)}: {balance}";
    }
  }

  public class Demo
  {
    static void Main(string[] args)
    {
      var ba = new BankAccount(100);
      var m1 = ba.Deposit(50); // 150
      var m2 = ba.Deposit(25); // 175
      WriteLine(ba);

      // restore to m1
      ba.Restore(m1);
      WriteLine(ba);

      ba.Restore(m2);
      WriteLine(ba);
    }
  }
}